src/Processors/Formats/Impl/Parquet/ParquetDataBuffer.h

```diff
@@ -0,0 +1,185 @@
+
+    template <typename TValue>
+    void ALWAYS_INLINE readValue(TValue & dst)
+    {
+        readBytes(&dst, sizeof(TValue));
+    }
+
+    void ALWAYS_INLINE readBytes(void * dst, size_t bytes)
+    {
+        checkAvaible(bytes);
+        std::copy(data, data + bytes, reinterpret_cast<Int8 *>(dst));
+        consume(bytes);
+    }
+
+    void ALWAYS_INLINE readDateTime64(DateTime64 & dst)
+    {
+        static const int max_scale_num = 9;
+        static const UInt64 pow10[max_scale_num + 1]
+            = {1000000000, 100000000, 10000000, 1000000, 100000, 10000, 1000, 100, 10, 1};
+        static const UInt64 spd = 60 * 60 * 24;
+        static const UInt64 scaled_day[max_scale_num + 1]
+            = {spd,
+               10 * spd,
+               100 * spd,
+               1000 * spd,
+               10000 * spd,
+               100000 * spd,
+               1000000 * spd,
+               10000000 * spd,
+               100000000 * spd,
+               1000000000 * spd};
+
+        parquet::Int96 tmp;
+        readValue(tmp);
+        auto decoded = parquet::DecodeInt96Timestamp(tmp);
+
+        uint64_t scaled_nano = decoded.nanoseconds / pow10[datetime64_scale];
+        dst = static_cast<Int64>(decoded.days_since_epoch * scaled_day[datetime64_scale] + scaled_nano);
+    }
+
+    /**
+     * This method should only be used to read string whose elements size is small.
+     * Because memcpySmallAllowReadWriteOverflow15 instead of memcpy is used according to ColumnString::indexImpl
+     */
+    void ALWAYS_INLINE readString(ColumnString & column, size_t cursor)
+    {
+        // refer to: PlainByteArrayDecoder::DecodeArrowDense in encoding.cc
+        //           deserializeBinarySSE2 in SerializationString.cpp
+        checkAvaible(4);
+        auto value_len = ::arrow::util::SafeLoadAs<Int32>(getArrowData());
+        if (unlikely(value_len < 0 || value_len > INT32_MAX - 4))
+        {
+            throw Exception(ErrorCodes::PARQUET_EXCEPTION, "Invalid or corrupted value_len '{}'", value_len);
+        }
+        consume(4);
+        checkAvaible(value_len);
+
+        auto chars_cursor = column.getChars().size();
+        column.getChars().resize(chars_cursor + value_len + 1);
+
+        memcpySmallAllowReadWriteOverflow15(&column.getChars()[chars_cursor], data, value_len);
+        column.getChars().back() = 0;
+
+        column.getOffsets().data()[cursor] = column.getChars().size();
+        consume(value_len);
+    }
+
+    template <is_over_big_decimal TDecimal>
+    void ALWAYS_INLINE readOverBigDecimal(TDecimal * out, Int32 elem_bytes_num)
+    {
+        using TArrowDecimal = typename ToArrowDecimal<TDecimal>::ArrowDecimal;
+
+        checkAvaible(elem_bytes_num);
+
+        // refer to: RawBytesToDecimalBytes in reader_internal.cc, Decimal128::FromBigEndian in decimal.cc
+        auto status = TArrowDecimal::FromBigEndian(getArrowData(), elem_bytes_num);
+        if (unlikely(!status.ok()))
```

Comment:
Indeed, `elem_bytes_num` may be not integral multiple of a word, it may be 18 bytes for example, so it's not such simple to read such big endian bytes as decimal. And we can find the logic in source code of [Decimal256::FromBigEndian](https://github.com/ClickHouse/arrow/blob/8f36d71d18587f1f315ec832f424183cb6519cbb/cpp/src/arrow/util/decimal.cc#L837 ).

About the status check, I can change it into an assert.